src/robot/can_adapter/adapter.py

```diff
@@ -0,0 +1,52 @@
+"""
+CAN adapter module.
+"""
+import asyncio
+from typing import Callable, Dict
+
+import can
+
+
+class CANAdapter:
+    """
+    This adapter abstracts away sending and receving messages on the CAN bus.
+    """
+
+    def __init__(self, can_bus: can.Bus):
+        self._can_bus = can_bus
+        self._notifier = can.Notifier(can_bus, [],
+                                      loop=asyncio.get_event_loop())
+
+    def stop(self) -> None:
+        """
+        Stop routing messages to the handlers. 
+        """
+        self._notifier.stop()
+
+    def send(self, arbitration_id: int, data: bytes) -> None:
+        """
+        Send a message on the bus.
+        """
+        self._can_bus.send(can.Message(arbitration_id=arbitration_id,
+                                       data=data))
+
+    def register_handlers(self,
+                          handlers: Dict[int, Callable[[bytes], None]]) -> None:
+        """
+        Register handlers to be called on receiving messages with a certain arbitration ID.
+        """
+        for arbitration_id, handler_func in handlers.items():
+            self.register_handler(arbitration_id, handler_func)
+
+    def register_handler(self, arbitration_id: int,
+                         handler: Callable[[bytes], None]) -> None:
+        """
+        Register a handler to be called on receiving messages with a certain arbitration ID.
+        """
+
+        def listener(msg: can.Message) -> None:
+            if msg.arbitration_id != arbitration_id:
+                return
+            handler(bytes(msg.data))
```

Comment:
To check : is msg.data exactly of the length of the message that was received?
In C you would get an array/pointer, and only use the "msg.dlc" first bytes of it. msg.dlc is the real payload size.
How does python-can handle payloads exactly?

Reply:
See https://python-can.readthedocs.io/en/master/message.html

> is msg.data exactly of the length of the message that was received?

Yes:
![image](https://user-images.githubusercontent.com/36550428/69259932-54590280-0bbf-11ea-805e-5fd03b250c11.png)

> msg.dlc is the real payload size.
> How does python-can handle payloads exactly?

dlc might not be the real payload size, see:

![image](https://user-images.githubusercontent.com/36550428/69259902-486d4080-0bbf-11ea-8a3b-b802d87951c4.png)